### ISSUE TITLE
[NLL] variable does not need to be mutable

### ISSUE THREAD

**Issue by kpp** [rust-lang/rust#51830]
UPD:
[minimal repro](https://play.rust-lang.org/?gist=96e87e9946fca3db5273c6ac87008af9&version=nightly&mode=debug) from @lqd :
```
#![feature(nll)]
#![allow(unused_variables)]
fn main() {
    let _ = Some(0).as_ref().map(|ref a| true);
}
```

---------------

I got this warning:

```
warning: variable does not need to be mutable
   --> src/toxcore/dht/server/mod.rs:363:32
    |
363 |         ping_map.retain(|&_pk, ref client|
    |                                ----^^^^^^
    |                                |
    |                                help: remove this `mut`
```

While I don't see any `mut` in the given code. The given patch fixed the warning but still I don't know why:

```diff
-        ping_map.retain(|&_pk, ref client|
+        ping_map.retain(|_pk, client|
```

You may try on: https://github.com/tox-rs/tox/commit/3d0ddf86c32abb94c94eb44ff1388bab06f50bd2 (you will have to enable NLL on your own).

Also seems like you don't highlight the correct site of the error in macros:

```
warning: variable does not need to be mutable
   --> src/toxcore/state_format/old.rs:173:9
    |
173 | /         do_gen!(buf,
174 | |             gen_le_u16!(0x0002) >>
175 | |             gen_slice!(SECTION_MAGIC) >>
176 | |             gen_le_u32!(DHT_MAGICAL as u32) >>
...   |
180 | |             gen_many_ref!(&self.0, |buf, node| PackedNode::to_bytes(node, buf))
181 | |         )
    | |_________^ help: remove this `mut`
    |
    = note: #[warn(unused_mut)] on by default
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

warning: variable does not need to be mutable
 --> <gen_many_ref macros>:3:32
  |
3 | Ok ( ( $ i , $ idx ) ) , | r , ref v | {
  |                                ----^
  |                                |
  |                                help: remove this `mut`
```

I would like you to highlight [gen_many_ref!](https://docs.rs/cookie-factory/0.2.4/src/cookie_factory/gen.rs.html#813-825).

**Comment by leonardo-m** [comment]
Dupe of #50897 ?

**Comment by kpp** [comment]
I tried to search for dups before I posted my issue. The difference with #50897 is that there is `mut` keyword in the code while my code does not have it.

**Comment by lqd** [comment]
I've had this one while trying to bootstrap rustc with NLL turned on, and had this [minimal repro](https://play.rust-lang.org/?gist=96e87e9946fca3db5273c6ac87008af9&version=nightly&mode=debug)

**Comment by nikomatsakis** [comment]
Tagging as NLL-deferred because #51918 has precedence
